### DISCUSSION TITLE
navigator.clipboard is undefined

### DISCUSSION THREAD

**Discussion by MemphisEx** [MemphisEx/Shopping_MERN#3] [Q&A] [answered]
Why is navigator.clipboard always undefined in the following snippet?

`
var clipboard = navigator.clipboard;
if (clipboard == undefined) {
    console.log('clipboard is undefined');
} else {
    clipboard.writeText('stuff to write').then(function() {
        console.log('Copied to clipboard successfully!');
    }, function() {
        console.error('Unable to write to clipboard. :-(');
    });
}`
More on the clipboard API can be found [here](https://www.w3.org/TR/clipboard-apis/#h-clipboard-writetext-data).

Chrome Version: 68.0.3440.106.

I'm sure this was working at some point, but no longer is. It's confusing because [this table](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API#Browser_Compatibility) suggests that the Clipboard API is implemented in Chrome (has been for some time), but [this table](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard#Browser_Compatibility) of specific API methods suggests that none of the methods of the API is supported??

**Comment by VladaWebDev** [comment, marked as answer]
Hi, you can write an all-in-one wrapper function.

- if in secure context (https) : use navigator clipboard api
- if not : use the 'out of viewport hidden text area' trick

// return a promise
function copyToClipboard(textToCopy) {
    // navigator clipboard api needs a secure context (https)
    if (navigator.clipboard && window.isSecureContext) {
        // navigator clipboard api method'
        return navigator.clipboard.writeText(textToCopy);
    } else {
        // text area method
        let textArea = document.createElement("textarea");
        textArea.value = textToCopy;
        // make the textarea out of viewport
        textArea.style.position = "fixed";
        textArea.style.left = "-999999px";
        textArea.style.top = "-999999px";
        document.body.appendChild(textArea);
        textArea.focus();
        textArea.select();
        return new Promise((res, rej) => {
            // here the magic happens
            document.execCommand('copy') ? res() : rej();
            textArea.remove();
        });
    }
}
use :

copyToClipboard("I'm going to the clipboard !")
    .then(() => console.log('text copied !'))
    .catch(() => console.log('error'));

**Reply by MemphisEx** [reply]
Thanks for your effort, @VladaWebDev
